ipa-core/src/seq_join.rs

```diff
@@ -667,4 +667,50 @@ mod test {
         let err = seq_try_join_all(active, (1..=3).map(f)).await.unwrap_err();
         assert_eq!(err, ERROR);
     }
+
+    /// This test demonstrates that forgetting the future returned by `parallel_join` is not safe and will cause
+    /// use-after-free safety error.
+    #[tokio::test(flavor = "multi_thread")]
+    #[cfg(feature = "multi-threading")]
+    #[ignore] // sanitizers will flag this test
+    async fn parallel_join_forget_is_not_safe() {
+        use futures::future::poll_immediate;
+
+        use crate::{seq_join::multi_thread::parallel_join, sync::Arc};
+
+        const N: usize = 24;
+        let borrow_from_me = vec![1, 2, 3];
+        let barrier1 = Arc::new(tokio::sync::Barrier::new(N + 1));
+        let barrier2 = Arc::new(tokio::sync::Barrier::new(N + 1));
+
+        let iterable = (0..N)
+            .map(|i| {
+                let borrowed = &borrow_from_me;
+                let b1 = barrier1.clone();
+                let b2 = barrier2.clone();
+                async move {
+                    b1.wait().await;
+                    for _ in 0..100 {
+                        if borrowed != &vec![1, 2, 3] {
```

Comment:
Can you deal with this differently, for instance, by passing in a weak reference that should remain live while the outside is operating.  That would be less prone to problems of scope.